docs/google-drive.md

```diff
@@ -0,0 +1,278 @@
+
+Our current system seems to largely accomplish these goals. We still
+occasionally trigger rate limits when onboarding a large group of people
+simultaneously, but we can generally still recover from this because of our
+just-in-time verification and initialization logic.
+
+To satisfy the first requirement, we grant the writer permission to "anyone" on
+each file. (This is the equivalent of the "Anyone on the internet with the link
+can edit" setting). This is the only permission we set on individual files and
+the only place we grant any full writer permissions. When a puzzle is deleted,
+we remove this permission to make the corresponding file read-only.
+
+Since all files for a hunt live in a shared folder, we grant permissions on that
+folder to any Google accounts linked to hunt members. Doing this once per folder
+instead of once per puzzle avoids needing to call the create permissions API
+once per puzzle per user, which would reliably result in rate limiting. Granting
+_any_ permission on a folder is sufficient to prevent Anonymous Animals on files
+in that folder. The "commenter" permission also results in edits being
+attributed to that user in the Drive Activity API (instead of the file's owner),
+while also preventing the user from being able to move or delete the files
+themselves.
+
+These permissions are granted by the `ensureHuntFolderPermission` function in
+`imports/server/gdrive.ts`, which in turn is called when a user is added to a
+new hunt (`addUserToHunt`), when they link a new Google account
+(`linkUserGoogleAccount`), or when they open a puzzle page (`ensureDocument`).
+
+Any time we successfully grant a user access to a hunt folder, we keep a flag
+record in the `FolderPermissions` model. This allows us to make permissions
+grants idempotent without needing additional API requests. This allows us to be
+resilient both to transient issues and rate limiting from Google's APIs and
+changes in state of the various circuit breaker feature flags covering the
+Google Drive integration, while also being proactive about ensuring the correct
+permissions are in place.
+
+## Activity tracking
+
+Edits on Google Drive files are a key indicator of solving activity on a puzzle,
+so we collect activity data using the [Drive Activity API][]. The server runs a
+loop in the background to periodically fetch activity data
+(`imports/server/gdriveActivityFetcher.ts`).
+
+The Drive Activity API lacks any way to request "records we haven't already
+seen", so instead we filter based on timestamp. We store the most recent
+timestamp we've observed in the `DriveActivityLatests` model (which only
+contains a single record). Activity is tracked in the `DocumentActivities`
+model, with a separate record per user so that we can deduplicate a single user
+acting across Google Drive, voice calls, and Jolly Roger chat in the same time
+bucket.
+
+As a note, we experimented with using Google Drive watches (both
```

Comment:
No action needed, I just wanted to say I especially love this paragraph detailing a negative result and the reasoning behind why we rejected that approach. 🙂